tools/integration-tests/tester/tests/testutil.go

```diff
@@ -619,3 +620,131 @@ func txMetadataStateEqual(t *testing.T, node *framework.Node, txID string, expIn
 	}
 	return true, metadata.GradeOfFinality
 }
+
+// IsBranchConfirmedOnAllPeers returns true if the branch is confirmed on all supplied nodes.
+func IsBranchConfirmedOnAllPeers(peers []*framework.Node, branchID string) bool {
+	for _, peer := range peers {
+		branch, err := peer.GetBranch(branchID)
+		if err != nil {
+			return false
+		}
+		if branch.GradeOfFinality != gof.High {
+			return false
+		}
+	}
+	return true
+}
+
+func findAttachmentMsg(peer *framework.Node, branchID string) (tip string, err error) {
+	branch, err := peer.GetBranch(branchID)
+	if err != nil {
+		return
+	}
+	attachments, err := peer.GetTransactionAttachments(branchID)
+	if err != nil {
+		return
+	}
+	children := attachments.MessageIDs
+	seen := make(map[string]struct{})
+	for len(children) > 0 {
```

Comment:
This loop over `children`, preventing to process a msgID twice is kinda hard to read. Hive.go has a datastructure exactly for cases like these, check it out https://github.com/iotaledger/hive.go/blob/5c76c0ccc6682e898bd45681ad5f3502b5f3ee50/datastructure/walker/walker.go

Could make the code significantly cleaner!